# Exercise 1

In [1]:
import pandas as pd

file_path = '/content/Report.csv'

# Load the CSV file into a Pandas DataFrame
df = pd.read_csv(file_path)

# Display the first 5 rows of the DataFrame
print("First 5 rows:")
display(df.head(5))

# Display the last 10 rows of the DataFrame
print("Last 10 rows:")
display(df.tail(10))

First 5 rows:


,id,Indeks,Prisustvo,Ispit1,Ispit2,Ispit1_popravni,Ispit2_popravni,UKUPNO,Ocjena
0,1,94-ST,10,9.4,7.5,12.08,13.73,35.81,7
1,2,77-ST,10,8.5,/,/,/,35.50,7
2,3,69-ST,10,17,/,/,/,39.50,9
3,4,79-ST,0,7.6,4.8,/,/,12.40,/
4,5,89-ST,10,10.6,12.5,12.58,/,35.08,6


Last 10 rows:


,id,Indeks,Prisustvo,Ispit1,Ispit2,Ispit1_popravni,Ispit2_popravni,UKUPNO,Ocjena
37,38,71-ST,10,/,/,/,/,42.16,8
38,39,54-ST,10,/,/,/,/,19.35,/
39,40,60-ST,10,10,/,/,/,38.50,6
40,41,48-ST,10,/,/,/,/,10.00,6
41,42,61-ST,10,/,/,/,/,23.12,/
42,43,1-ST,10,/,/,/,/,10.00,/
43,44,88-ST,10,/,/,/,/,37.00,7
44,45,75-ST,10,/,/,9.83,13.5,33.33,6
45,46,67-ST,10,16.5,/,/,/,39.50,7
46,47,81-ST,10,15.7,17.3,/,/,43.00,8


In [2]:
# Display only the columns related to regular partial exams (Ispit1 and Ispit2)
print("Regular Partial Exams:")
display(df[['Ispit1', 'Ispit2']])

Regular Partial Exams:


,Ispit1,Ispit2
0,9.4,7.5
1,8.5,/
2,17,/
3,7.6,4.8
4,10.6,12.5
5,/,/
6,/,12.5
7,/,/
8,/,/
9,/,/


In [3]:
# Display students who lost attendance
print("Students who lost attendance:")
display(df[df['Prisustvo'] == 0])

Students who lost attendance:


,id,Indeks,Prisustvo,Ispit1,Ispit2,Ispit1_popravni,Ispit2_popravni,UKUPNO,Ocjena
3,4,79-ST,0,7.6,4.8,/,/,12.40,/
5,6,78-ST,0,/,/,13.08,12.83,25.91,6
8,9,83-ST,0,/,/,/,/,0.00,/
22,23,74-ST,0,10.7,/,14.58,11.7,27.08,/
28,29,95-ST,0,10.8,12.5,7.25,/,23.50,/
31,32,41-ST,0,/,/,/,/,5.87,/
32,33,17-ST,0,7.5,1.7,2.33,/,9.20,/
33,34,82-ST,0,10,1.5,/,/,16.00,/
34,35,15-ST,0,/,/,/,/,0.00,/


In [6]:
# Clean and convert 'Ocjena' column to numeric
df['Ocjena'] = df['Ocjena'].astype(str).str.strip()  # Remove extra spaces
df['Ocjena'] = pd.to_numeric(df['Ocjena'], errors='coerce')  # Convert to numeric

# Display index, total points, and grade for students with grade 8 or higher
print("Students with grade 8 or higher:")
display(df[df['Ocjena'] >= 8][['Indeks', 'UKUPNO', 'Ocjena']])

Students with grade 8 or higher:


,Indeks,UKUPNO,Ocjena
2,69-ST,39.50,9.0
6,93-ST,35.17,8.0
7,68-ST,50.00,9.0
9,92-ST,42.06,9.0
10,65-ST,43.19,8.0
12,90-ST,39.90,8.0
13,97-ST,41.05,9.0
14,84-ST,43.56,9.0
16,91-ST,43.48,9.0
19,96-ST,41.40,8.0


In [9]:
import numpy as np

# Replace missing values (/) with np.nan
df.replace('/', np.nan, inplace=True)

print("First 5 rows:")
display(df.head(5))

First 5 rows:


,id,Indeks,Prisustvo,Ispit1,Ispit2,Ispit1_popravni,Ispit2_popravni,UKUPNO,Ocjena
0,1,94-ST,10,9.4,7.5,12.08,13.73,35.81,7.0
1,2,77-ST,10,8.5,NaN,NaN,NaN,35.50,7.0
2,3,69-ST,10,17,NaN,NaN,NaN,39.50,9.0
3,4,79-ST,0,7.6,4.8,NaN,NaN,12.40,NaN
4,5,89-ST,10,10.6,12.5,12.58,NaN,35.08,6.0


In [10]:
# Remove students who do not have a recorded grade
df = df.dropna(subset=['Ocjena'])

print("First 5 rows:")
display(df.head(5))

First 5 rows:


,id,Indeks,Prisustvo,Ispit1,Ispit2,Ispit1_popravni,Ispit2_popravni,UKUPNO,Ocjena
0,1,94-ST,10,9.4,7.5,12.08,13.73,35.81,7.0
1,2,77-ST,10,8.5,NaN,NaN,NaN,35.50,7.0
2,3,69-ST,10,17,NaN,NaN,NaN,39.50,9.0
4,5,89-ST,10,10.6,12.5,12.58,NaN,35.08,6.0
5,6,78-ST,0,NaN,NaN,13.08,12.83,25.91,6.0


In [12]:
# Convert 'Ispit1' and 'Ispit1_popravni' columns to numeric values
df['Ispit1'] = pd.to_numeric(df['Ispit1'], errors='coerce')
df['Ispit1_popravni'] = pd.to_numeric(df['Ispit1_popravni'], errors='coerce')

# Create a new column 'Ispit1_final' with the better result between Ispit1 and Ispit1_popravni
df['Ispit1_final'] = df[['Ispit1', 'Ispit1_popravni']].max(axis=1, skipna=True)

# Display the updated DataFrame with 'Ispit1_final' column
print("Updated DataFrame with Ispit1_final:")
display(df[['Indeks', 'Ispit1', 'Ispit1_popravni', 'Ispit1_final']])

Updated DataFrame with Ispit1_final:


,Indeks,Ispit1,Ispit1_popravni,Ispit1_final
0,94-ST,9.4,12.08,12.08
1,77-ST,8.5,NaN,8.50
2,69-ST,17.0,NaN,17.00
4,89-ST,10.6,12.58,12.58
5,78-ST,NaN,13.08,13.08
6,93-ST,NaN,12.67,12.67
7,68-ST,NaN,NaN,NaN
9,92-ST,NaN,NaN,NaN
10,65-ST,NaN,NaN,NaN
11,70-ST,13.0,NaN,13.00


In [13]:
# Convert 'Ispit2' and 'Ispit2_popravni' columns to numeric values
df['Ispit2'] = pd.to_numeric(df['Ispit2'], errors='coerce')
df['Ispit2_popravni'] = pd.to_numeric(df['Ispit2_popravni'], errors='coerce')

# Create a new column 'Ispit2_final' with the better result between Ispit2 and Ispit2_popravni
df['Ispit2_final'] = df[['Ispit2', 'Ispit2_popravni']].max(axis=1, skipna=True)

# Display the updated DataFrame with 'Ispit2_final' column
print("Updated DataFrame with Ispit2_final:")
display(df[['Indeks', 'Ispit2', 'Ispit2_popravni', 'Ispit2_final']])

Updated DataFrame with Ispit2_final:


,Indeks,Ispit2,Ispit2_popravni,Ispit2_final
0,94-ST,7.5,13.73,13.73
1,77-ST,NaN,NaN,NaN
2,69-ST,NaN,NaN,NaN
4,89-ST,12.5,NaN,12.50
5,78-ST,NaN,12.83,12.83
6,93-ST,12.5,NaN,12.50
7,68-ST,NaN,NaN,NaN
9,92-ST,NaN,NaN,NaN
10,65-ST,NaN,NaN,NaN
11,70-ST,7.8,16.30,16.30


In [14]:
# Drop the old columns for the exams: 'Ispit1', 'Ispit1_popravni', 'Ispit2', 'Ispit2_popravni'
df.drop(columns=['Ispit1', 'Ispit1_popravni', 'Ispit2', 'Ispit2_popravni'], inplace=True)

# Display the updated DataFrame without the old exam columns
print("Updated DataFrame without the old exam columns:")
display(df)

Updated DataFrame without the old exam columns:


,id,Indeks,Prisustvo,UKUPNO,Ocjena,Ispit1_final,Ispit2_final
0,1,94-ST,10,35.81,7.0,12.08,13.73
1,2,77-ST,10,35.50,7.0,8.50,NaN
2,3,69-ST,10,39.50,9.0,17.00,NaN
4,5,89-ST,10,35.08,6.0,12.58,12.50
5,6,78-ST,0,25.91,6.0,13.08,12.83
6,7,93-ST,10,35.17,8.0,12.67,12.50
7,8,68-ST,10,50.00,9.0,NaN,NaN
9,10,92-ST,10,42.06,9.0,NaN,NaN
10,11,65-ST,10,43.19,8.0,NaN,NaN
11,12,70-ST,10,39.30,7.0,13.00,16.30


In [15]:
# Save the modified DataFrame to a CSV file with semicolon separator
output_file_path = '/content/Report_modified.csv'
df.to_csv(output_file_path, sep=';', index=False)

'/content/Report_modified.csv'

In [16]:
# Save the modified DataFrame as a Pickle file
pickle_file_path = '/content/Report_modified_pickle.p'
df.to_pickle(pickle_file_path)

'/content/Report_modified_pickle.p'

# Exercise 2

In [19]:
import pandas as pd

file_path = '/content/Report.csv'

# Load the CSV file into a Pandas DataFrame
df = pd.read_csv(file_path)

# Display the first few rows to confirm the DataFrame is loaded
df.head()

,id,Indeks,Prisustvo,Ispit1,Ispit2,Ispit1_popravni,Ispit2_popravni,UKUPNO,Ocjena
0,1,94-ST,10,9.4,7.5,12.08,13.73,35.81,7
1,2,77-ST,10,8.5,/,/,/,35.50,7
2,3,69-ST,10,17,/,/,/,39.50,9
3,4,79-ST,0,7.6,4.8,/,/,12.40,/
4,5,89-ST,10,10.6,12.5,12.58,/,35.08,6


In [20]:
from sklearn.impute import SimpleImputer
import numpy as np

# Replace '/' with NaN in the relevant columns
df[['Ispit1', 'Ispit2']] = df[['Ispit1', 'Ispit2']].replace('/', np.nan)

# Create an imputer with median strategy
imputer = SimpleImputer(strategy='median')

# Apply the imputer to the columns with potential missing values
df[['Ispit1', 'Ispit2']] = imputer.fit_transform(df[['Ispit1', 'Ispit2']])

# Display the DataFrame after imputation
df.head()


,id,Indeks,Prisustvo,Ispit1,Ispit2,Ispit1_popravni,Ispit2_popravni,UKUPNO,Ocjena
0,1,94-ST,10,9.4,7.5,12.08,13.73,35.81,7
1,2,77-ST,10,8.5,8.7,/,/,35.50,7
2,3,69-ST,10,17.0,8.7,/,/,39.50,9
3,4,79-ST,0,7.6,4.8,/,/,12.40,/
4,5,89-ST,10,10.6,12.5,12.58,/,35.08,6


In [21]:
# Replace '/' with NaN in the relevant columns
df[['Ispit1_popravni', 'Ispit2_popravni']] = df[['Ispit1_popravni', 'Ispit2_popravni']].replace('/', np.nan)

# Create an imputer with median strategy
imputer = SimpleImputer(strategy='mean')

# Apply the imputer to the columns with potential missing values
df[['Ispit1_popravni', 'Ispit2_popravni']] = imputer.fit_transform(df[['Ispit1_popravni', 'Ispit2_popravni']])

# Display the DataFrame after imputation
df.head()

,id,Indeks,Prisustvo,Ispit1,Ispit2,Ispit1_popravni,Ispit2_popravni,UKUPNO,Ocjena
0,1,94-ST,10,9.4,7.5,12.080,13.730000,35.81,7
1,2,77-ST,10,8.5,8.7,10.657,14.662727,35.50,7
2,3,69-ST,10,17.0,8.7,10.657,14.662727,39.50,9
3,4,79-ST,0,7.6,4.8,10.657,14.662727,12.40,/
4,5,89-ST,10,10.6,12.5,12.580,14.662727,35.08,6


In [22]:
from sklearn.preprocessing import StandardScaler

# Initialize the StandardScaler
scaler = StandardScaler()

# Select the columns representing the regular exams
exam_columns = ['Ispit1', 'Ispit2']

# Apply Z-score normalization to the selected columns
df[exam_columns] = scaler.fit_transform(df[exam_columns])

# Display the DataFrame after Z-score normalization
df.head()

,id,Indeks,Prisustvo,Ispit1,Ispit2,Ispit1_popravni,Ispit2_popravni,UKUPNO,Ocjena
0,1,94-ST,10,-1.561766,-0.555596,12.080,13.730000,35.81,7
1,2,77-ST,10,-2.029964,-0.087900,10.657,14.662727,35.50,7
2,3,69-ST,10,2.391904,-0.087900,10.657,14.662727,39.50,9
3,4,79-ST,0,-2.498161,-1.607912,10.657,14.662727,12.40,/
4,5,89-ST,10,-0.937502,1.393136,12.580,14.662727,35.08,6


In [23]:
from sklearn.preprocessing import MinMaxScaler

# Initialize the MinMaxScaler
scaler = MinMaxScaler()

# Select the columns representing the retake exams
retake_columns = ['Ispit1_popravni', 'Ispit2_popravni']

# Apply Min-Max normalization to the selected columns
df[retake_columns] = scaler.fit_transform(df[retake_columns])

# Display the DataFrame after Min-Max normalization
df.head()

,id,Indeks,Prisustvo,Ispit1,Ispit2,Ispit1_popravni,Ispit2_popravni,UKUPNO,Ocjena
0,1,94-ST,10,-1.561766,-0.555596,0.795918,0.401146,35.81,7
1,2,77-ST,10,-2.029964,-0.087900,0.679755,0.490232,35.50,7
2,3,69-ST,10,2.391904,-0.087900,0.679755,0.490232,39.50,9
3,4,79-ST,0,-2.498161,-1.607912,0.679755,0.490232,12.40,/
4,5,89-ST,10,-0.937502,1.393136,0.836735,0.490232,35.08,6


In [24]:
# Replace all remaining missing values in the entire DataFrame with zeros
df = df.fillna(0)

# Display the DataFrame after filling missing values with zeros
df.head()

,id,Indeks,Prisustvo,Ispit1,Ispit2,Ispit1_popravni,Ispit2_popravni,UKUPNO,Ocjena
0,1,94-ST,10,-1.561766,-0.555596,0.795918,0.401146,35.81,7
1,2,77-ST,10,-2.029964,-0.087900,0.679755,0.490232,35.50,7
2,3,69-ST,10,2.391904,-0.087900,0.679755,0.490232,39.50,9
3,4,79-ST,0,-2.498161,-1.607912,0.679755,0.490232,12.40,/
4,5,89-ST,10,-0.937502,1.393136,0.836735,0.490232,35.08,6


In [27]:
# Extract the 'Ocjena' column into a separate variable
ocjena = df['Ocjena']

# Drop the 'Ocjena' column from the original DataFrame
df = df.drop(columns=['Ocjena'])

# Display the modified DataFrame and the extracted 'Ocjena' variable
print(df.head())
print("\n")
print(ocjena.head())

   id Indeks  Prisustvo    Ispit1    Ispit2  Ispit1_popravni  Ispit2_popravni  \
0   1  94-ST         10 -1.561766 -0.555596         0.795918         0.401146   
1   2  77-ST         10 -2.029964 -0.087900         0.679755         0.490232   
2   3  69-ST         10  2.391904 -0.087900         0.679755         0.490232   
3   4  79-ST          0 -2.498161 -1.607912         0.679755         0.490232   
4   5  89-ST         10 -0.937502  1.393136         0.836735         0.490232   

   UKUPNO  
0   35.81  
1   35.50  
2   39.50  
3   12.40  
4   35.08  


0    7
1    7
2    9
3    /
4    6
Name: Ocjena, dtype: object


In [28]:
# Convert the original DataFrame (without 'Ocjena' column) to a NumPy array
df_numpy = df.to_numpy()

# Convert the 'Ocjena' variable to a NumPy array
ocjena_numpy = ocjena.to_numpy()

# Display the results
print(df_numpy)
print(ocjena_numpy)

[[1 '94-ST' 10 -1.5617659452228656 -0.5555960491403172 0.7959183673469388
  0.4011461318051578 35.81]
 [2 '77-ST' 10 -2.029963673663172 -0.08790027046100546 0.6797551020408164
  0.4902318312060433 35.5]
 [3 '69-ST' 10 2.3919037616063856 -0.08790027046100546 0.6797551020408164
  0.4902318312060433 39.5]
 [4 '79-ST' 0 -2.498161402103478 -1.6079115511687696 0.6797551020408164
  0.4902318312060433 12.4]
 [5 '89-ST' 10 -0.937502307302458 1.3931363620234831 0.836734693877551
  0.4902318312060433 35.08]
 [6 '78-ST' 0 0.050915119404855075 -0.08790027046100546
  0.8775510204081632 0.31518624641833826 25.91]
 [7 '93-ST' 10 0.050915119404855075 1.3931363620234831 0.8440816326530612
  0.4902318312060433 35.17]
 [8 '68-ST' 10 0.050915119404855075 -0.08790027046100546
  0.6797551020408164 0.4902318312060433 50.0]
 [9 '83-ST' 0 0.050915119404855075 -0.08790027046100546
  0.6797551020408164 0.4902318312060433 0.0]
 [10 '92-ST' 10 0.050915119404855075 -0.08790027046100546
  0.6797551020408164 0.4902318

In [29]:
from sklearn.model_selection import train_test_split

# Split the data into training and test sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(df_numpy, ocjena_numpy, test_size=0.2, random_state=42)

# Display the shapes of the resulting sets to confirm the split
print("Training features shape:", X_train.shape)
print("Test features shape:", X_test.shape)
print("Training labels shape:", y_train.shape)
print("Test labels shape:", y_test.shape)

Training features shape: (37, 8)
Test features shape: (10, 8)
Training labels shape: (37,)
Test labels shape: (10,)


# Exercise 3

In [30]:
from sklearn.datasets import load_iris

# Load the Iris dataset
iris = load_iris()

# Features (attributes)
X = iris.data

# Labels (target classes)
y = iris.target

# Feature names
feature_names = iris.feature_names

# Target names (class labels)
target_names = iris.target_names

# Print feature and target names
print("Feature names:", feature_names)
print("Target names:", target_names)

# Print the first 5 rows of the feature matrix
print("\nFirst 5 rows of X:\n", X[:5])

Feature names: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
Target names: ['setosa' 'versicolor' 'virginica']

First 5 rows of X:
 [[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]


In [31]:
from sklearn.model_selection import train_test_split

# Split the dataset into training and test sets
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=1
)

# Print the shapes of the resulting splits
print(X_train.shape)
print(X_test.shape)

print(y_train.shape)
print(y_test.shape)

(105, 4)
(45, 4)
(105,)
(45,)


New Shapes: The training set will have 105 samples, and the test set will have 45 samples.

Size of Dataset: The Iris dataset is relatively small, with only 150 samples. While it’s perfect for learning and small-scale testing, it might not be large enough for complex machine learning tasks that require large datasets.

In [32]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics

# Initialize the k-NN classifier with k=3
classifier_knn = KNeighborsClassifier(n_neighbors=3)

# Train the classifier using the training data
classifier_knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [33]:
y_pred = classifier_knn.predict(X_test)

print("Accuracy:", metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.9777777777777777


The accuracy of 97.78% is excellent. This means the k-NN model is performing very well on the Iris dataset, correctly classifying almost 98% of the test samples. For this basic classification task, this level of accuracy is considered very good.

**Is this accuracy satisfactory?**

This accuracy is highly satisfactory for a simple classification problem like Iris. It indicates that the model is learning effectively from the training data and generalizing well to the test data.

**Is the algorithm applicable in the real world?**

Given that the Iris dataset is small and clean, k-NN works well here. For simple, small-scale classification tasks like flower classification, k-NN is certainly applicable. However, for more complex or larger datasets, we might want to explore other algorithms to handle scalability and noise more effectively.

In [34]:
# Test different values of k
k_values = [1, 3, 5, 7, 9]  # Set of k values to test
accuracies = []

for k in k_values:
    # Initialize the classifier with the current value of k
    classifier_knn = KNeighborsClassifier(n_neighbors=k)

    # Fit the model on the training data
    classifier_knn.fit(X_train, y_train)

    # Make predictions on the test set
    y_pred = classifier_knn.predict(X_test)

    # Calculate accuracy
    accuracy = metrics.accuracy_score(y_test, y_pred)
    accuracies.append((k, accuracy))
    print(f"Accuracy for k={k}: {accuracy}")

Accuracy for k=1: 0.9777777777777777
Accuracy for k=3: 0.9777777777777777
Accuracy for k=5: 0.9777777777777777
Accuracy for k=7: 0.9777777777777777
Accuracy for k=9: 0.9777777777777777


It looks like the accuracy remains constant for all values of k (1, 3, 5, 7, and 9), all yielding an accuracy of 97.78%.

This suggests that, for this dataset, the choice of k in this range doesn't significantly impact the performance of the model. The k-NN algorithm seems to be stable in this case, and the accuracy remains consistent even when we change the number of neighbors.

Normally, you might expect the accuracy to change with varying k values, especially if k is very small or very large. However, the consistency here indicates that the Iris dataset is relatively simple and clean, so it doesn't require fine-tuning of k for good performance.

**What if k is equal to the number of test samples?**

If k were set to the number of test samples, the model would classify all test samples into one class, resulting in poor performance (around 33% accuracy for a 3-class problem like Iris).

# Exercise 4

In [35]:
from keras.datasets import mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


In [36]:
# Get the shape of the arrays
print(f"Shape of x_train: {x_train.shape}")
print(f"Shape of y_train: {y_train.shape}")
print(f"Shape of x_test: {x_test.shape}")
print(f"Shape of y_test: {y_test.shape}")

# Calculate memory usage
x_train_memory = x_train.nbytes / (1024 ** 2)  # Convert to MB
y_train_memory = y_train.nbytes / 1024  # Convert to KB
x_test_memory = x_test.nbytes / (1024 ** 2)  # Convert to MB
y_test_memory = y_test.nbytes / 1024  # Convert to KB

# Print memory usage
print(f"Memory usage of x_train: {x_train_memory:.2f} MB")
print(f"Memory usage of y_train: {y_train_memory:.2f} KB")
print(f"Memory usage of x_test: {x_test_memory:.2f} MB")
print(f"Memory usage of y_test: {y_test_memory:.2f} KB")

Shape of x_train: (60000, 28, 28)
Shape of y_train: (60000,)
Shape of x_test: (10000, 28, 28)
Shape of y_test: (10000,)
Memory usage of x_train: 44.86 MB
Memory usage of y_train: 58.59 KB
Memory usage of x_test: 7.48 MB
Memory usage of y_test: 9.77 KB


In [37]:
train_images = x_train.reshape((x_train.shape[0], 28 * 28))
train_images = train_images.astype('float32') / 255

test_images = x_test.reshape((x_test.shape[0], 28*28))
test_images = test_images.astype('float32') / 255

The reshape function is used to change the shape of the input data without changing the data itself. In this case:

*   x_train.shape[0] refers to the number of training samples (images) in the dataset.
*   28 * 28 is the flattened size of each image, because each MNIST image is originally 28x28 pixels. When reshaped, this will transform each 28x28 image into a one-dimensional array of length 784 (28 multiplied by 28).

This reshaping operation flattens the 28x28 images into a single array of 784 elements for each image. It allows the neural network to process the data as a vector rather than a matrix.



The astype function changes the data type of the array elements. In this case:

*   astype('float32') converts the pixel values to the float32 data type. This is important because most neural network frameworks (like Keras) expect floating-point numbers rather than integers, and it ensures that the numbers can represent decimal values.
*   / 255 scales the pixel values from the original range of 0 to 255 to a range between 0 and 1. This normalization is important for neural networks because it helps the model converge more quickly and ensures that all input values are on the same scale.



In [38]:
# Memory usage after reshaping and converting to float32
train_images_memory = train_images.nbytes / (1024 ** 2)  # Convert to MB
test_images_memory = test_images.nbytes / (1024 ** 2)  # Convert to MB

print(f"Memory usage of train_images: {train_images_memory:.2f} MB")
print(f"Memory usage of test_images: {test_images_memory:.2f} MB")

Memory usage of train_images: 179.44 MB
Memory usage of test_images: 29.91 MB


In [40]:
from keras import models
from keras import layers

network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
network.add(layers.Dense(10, activation='softmax'))

In [42]:
network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])


In [43]:
from keras.utils import to_categorical

train_labels = to_categorical(y_train)
test_labels = to_categorical(y_test)

In [44]:
network.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 5s 9ms/step - accuracy: 0.8751 - loss: 0.4359
Epoch 2/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 5s 9ms/step - accuracy: 0.9669 - loss: 0.1149
Epoch 3/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 6s 10ms/step - accuracy: 0.9782 - loss: 0.0739
Epoch 4/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 4s 9ms/step - accuracy: 0.9842 - loss: 0.0529
Epoch 5/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 5s 12ms/step - accuracy: 0.9887 - loss: 0.0383


In [46]:
test_loss, test_acc = network.evaluate(test_images, test_labels)
print('Accuracy for the test dataset:', test_acc)

313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9743 - loss: 0.0800
Accuracy for the test dataset: 0.9783999919891357


The accuracy on the test set is 97.84%.

The accuracy on the training set is 97.43%.

The difference between the training and test accuracy is very small, which means the model generalizes well and is not overly specific to the training data.